File: mixin-base/src/main/java/net/yakclient/mixin/base/internal/bytecode/TreeMixinCV.java
package net.yakclient.mixin.base.internal.bytecode;

import net.yakclient.mixin.base.internal.methodadapter.MixinPatternMatcher;
import org.objectweb.asm.ClassVisitor;
import org.objectweb.asm.MethodVisitor;
import org.objectweb.asm.tree.ClassNode;

import java.util.Map;
import java.util.Queue;

public class TreeMixinCV extends MixinCV {
    private final ClassVisitor parent;

    public TreeMixinCV(ClassVisitor parent, Map<String, Queue<QualifiedInstruction>> injectors) {
        super(new ClassNode(), injectors);
        this.parent = parent;
    }

    @Override
    public MethodVisitor visitMethod(int access, String name, String desc, String signature, String[] exceptions) {
        final MethodVisitor visitor = super.visitMethod(access, name, desc, signature, exceptions);

        final var qualifiedName = name + desc;
        if (this.hasInjection(qualifiedName)) {
            final Queue<QualifiedInstruction> instructions = this.getInjection(qualifiedName);
            final MixinPatternMatcher.TreeMixinPatternMatcherBuilder builder = MixinPatternMatcher.treePatternMatcher();
            for (var insn : instructions) {
                builder.addNode(insn.getInjectionType(), insn.getInsn(), insn.getPriority());
            }
            return builder.build(visitor);
        }

        return visitor;
    }

    @Override
    public void visitEnd() {
        ((ClassNode) this.cv).accept(this.parent);
    }
}

File: mixin-base/src/main/kotlin/net/yakclient/mixin/base/internal/bytecode/TreeMixinCV.kt
package net.yakclient.mixin.base.internal.bytecode

import net.yakclient.mixin.base.internal.methodadapter.MixinPatternMatcher
import net.yakclient.mixin.base.internal.methodadapter.PriorityMatcher
import net.yakclient.mixin.base.internal.methodadapter.tree.MethodProxyAdapter
import net.yakclient.mixin.base.internal.methodadapter.tree.TreeMethodProxy
import net.yakclient.mixin.base.internal.methodadapter.tree.TreeMixinPatternMatcher
import org.objectweb.asm.ClassVisitor
import org.objectweb.asm.MethodVisitor
import org.objectweb.asm.tree.ClassNode
import java.util.*

class TreeMixinCV(
    private val parent: ClassVisitor,
    injectors: Map<String, Queue<QualifiedInstruction>>
) : MixinCV(
    ClassNode(), injectors
) {
    override fun visitMethod(
        access: Int,
        name: String,
        desc: String,
        signature: String?,
        exceptions: Array<String>?
    ): MethodVisitor {
        val visitor = super.visitMethod(access, name, desc, signature, exceptions)
        val qualifiedName = name + desc
        if (hasInjection(qualifiedName)) {
            val instructions = getInjection(qualifiedName)
            val nodes = PriorityQueue<PriorityMatcher<TreeMixinPatternMatcher>>();
            for (insn in instructions) {
               nodes.add(MixinPatternMatcher.createTreeNode(insn.injectionType, insn.insn, insn.priority));
            }
            return   MethodProxyAdapter(TreeMethodProxy(nodes), visitor)
        }
        return visitor
    }

    override fun visitEnd() {
        (cv as ClassNode).accept(parent)
    }
}